pageserver/src/tenant.rs

```diff
@@ -1399,16 +1462,41 @@ impl Tenant {
     }
 
     pub(crate) async fn apply_timeline_archival_config(
-        &self,
+        self: &Arc<Self>,
         timeline_id: TimelineId,
         state: TimelineArchivalState,
+        ctx: RequestContext,
     ) -> Result<(), TimelineArchivalError> {
         info!("setting timeline archival config");
-        let timeline = {
+        let timeline_or_unarchive_offloaded = 'outer: {
             let timelines = self.timelines.lock().unwrap();
 
             let Some(timeline) = timelines.get(&timeline_id) else {
-                return Err(TimelineArchivalError::NotFound);
+                let offloaded_timelines = self.timelines_offloaded.lock().unwrap();
+                let Some(offloaded) = offloaded_timelines.get(&timeline_id) else {
+                    return Err(TimelineArchivalError::NotFound);
+                };
+                if state == TimelineArchivalState::Archived {
+                    // It's offloaded, so nothing to do for archival
+                    return Ok(());
+                }
+                if let Some(ancestor_timeline_id) = offloaded.ancestor_timeline_id {
+                    let has_archived_parent =
+                        if let Some(ancestor_timeline) = timelines.get(&timeline_id) {
+                            ancestor_timeline.is_archived() == Some(true)
+                        } else if offloaded_timelines.contains_key(&ancestor_timeline_id) {
+                            true
+                        } else {
+                            error!("ancestor timeline {ancestor_timeline_id} not found");
```

Comment:
Can we debug_assert!() against this case?  It should be impossible, right?

Reply:
There is an extremely unlikely case that the timeline is created parallel to the deletion of the ancestor, so the ancestor is removed from the list of timelines and then the timeline is inserted into it.

We do have a check in the deletion code that the timeline hasn't grown any children, but of course it still can be in the process of adding children. Or maybe there is something in timeline creation preventing that (say some code that, while holding the lock, ensures the parent is part of the list of timelines and then inserting the child), idk.

If that situation is possible, the timeline still needs to become offloaded or such. But that is even more unlikely. Certainly violation of invariants, so it merits a `debug_assert`.